domain/cloudimagemetadata/modelmigration/export_test.go

```diff
@@ -0,0 +1,126 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package modelmigration
+
+import (
+	"context"
+	"errors"
+
+	"github.com/juju/clock"
+	"github.com/juju/description/v8"
+	jc "github.com/juju/testing/checkers"
+	"go.uber.org/mock/gomock"
+	gc "gopkg.in/check.v1"
+
+	"github.com/juju/juju/domain/cloudimagemetadata"
+	loggertesting "github.com/juju/juju/internal/logger/testing"
+)
+
+type exportSuite struct {
+	coordinator *MockCoordinator
+	service     *MockExportService
+	clock       clock.Clock
+}
+
+var _ = gc.Suite(&exportSuite{})
+
+func (s *exportSuite) setupMocks(c *gc.C) *gomock.Controller {
+	ctrl := gomock.NewController(c)
+
+	s.coordinator = NewMockCoordinator(ctrl)
+	s.service = NewMockExportService(ctrl)
+	s.clock = clock.WallClock
+
+	return ctrl
+}
+
+func (s *exportSuite) newExportOperation(c *gc.C) *exportOperation {
+	return &exportOperation{
+		service: s.service,
+		logger:  loggertesting.WrapCheckLog(c),
+		clock:   s.clock,
+	}
+}
+
+// TestRegisterExport tests the registration of export operations with the coordinator.
+func (s *exportSuite) TestRegisterExport(c *gc.C) {
+	defer s.setupMocks(c).Finish()
+
+	s.coordinator.EXPECT().Add(gomock.Any())
+
+	RegisterExport(s.coordinator, loggertesting.WrapCheckLog(c))
+}
+
+// TestExport verifies the export of cloud image metadata to the model. It creates some metadata with different values
+// and check that all of them are added to the model.
+func (s *exportSuite) TestExport(c *gc.C) {
+	defer s.setupMocks(c).Finish()
+
+	// Arrange
+	creationTime := s.clock.Now()
+	defaultAttr := cloudimagemetadata.MetadataAttributes{
+		Stream:          "stream",
+		Region:          "region",
+		Version:         "version",
+		Arch:            "arch",
+		VirtType:        "virtType",
+		RootStorageType: "rootStorageType",
+		RootStorageSize: ptr(uint64(12)),
+		Source:          "custom",
+	}
+	attr1, attr2, attr3 := defaultAttr, defaultAttr, defaultAttr
+	attr2.Stream = "stream2"
+	attr3.RootStorageSize = nil
+
+	expected := []cloudimagemetadata.Metadata{{
+		MetadataAttributes: attr1,
+		Priority:           41,
+		ImageID:            "attr1",
+		CreationTime:       creationTime,
+	}, {
+		MetadataAttributes: attr2,
+		Priority:           42,
+		ImageID:            "attr2",
+		CreationTime:       creationTime,
+	}, {
+		MetadataAttributes: attr3,
+		Priority:           43,
+		ImageID:            "attr3",
+		CreationTime:       creationTime,
+	}}
+	dst := description.NewModel(description.ModelArgs{})
+	s.service.EXPECT().AllCloudImageMetadata(gomock.Any()).Return(expected, nil)
+
+	// Act
+	op := s.newExportOperation(c)
+	err := op.Execute(context.Background(), dst)
+	c.Assert(err, jc.ErrorIsNil)
+
+	// Assert
+	actualCloudMetadata := dst.CloudImageMetadata()
+	obtained := transformMetadataFromDescriptionToDomain(actualCloudMetadata)
```

Comment:
This is a matter of opinion really, but to me it would be better to explicitly declare the expected data rather than generate it here so that its really obvious to someone looking at the test.

By generating it from the input, if you added an extra field or something it wouldn't be so obvious that it was missing from the tests.

Reply:
It don't generate anything from input. That's transformation code from one structure to another, in order to be able to have a simpler assert, instead of field-to-field assert for each 3 Metadata (~40 lines).

However we can argue about the readability. If some field are missing, what is less costly to update/analyze:

- a single assert with transformation code
- ~40 asserts

For me it is close enough to not have a clear opinion, but the former don't override my brain by many one-to-one checks when i read the test without having to fix it :)

Reply:
I guess for me the benefit would be being able to look on inspection and just see if something was there or not in the expected data. But if writing out the expected data would generate too much noise this benefit goes away.

Reply:
Yeah don't think it worth it.